src/Microsoft.AspNetCore.Mvc.ApiExplorer/ApiResponseTypeProvider.cs

```diff
@@ -0,0 +1,328 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Formatters;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.ApiExplorer
+{
+    internal class ApiResponseTypeProvider
+    {
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+        private readonly IActionResultTypeMapper _mapper;
+        private readonly MvcOptions _mvcOptions;
+
+        public ApiResponseTypeProvider(
+            IModelMetadataProvider modelMetadataProvider,
+            IActionResultTypeMapper mapper,
+            MvcOptions mvcOptions)
+        {
+            _modelMetadataProvider = modelMetadataProvider;
+            _mapper = mapper;
+            _mvcOptions = mvcOptions;
+        }
+
+        public IList<ApiResponseType> GetApiResponseTypes(ControllerActionDescriptor action)
```

Comment:
Most of this, with the exception of the convention stuff, is moved from `DefaultApiDescriptionProvider.cs`. It was getting hard to manage that type